### PR TITLE
Feature/source openid

### PR DESCRIPTION
**Pull request by whitx**
# Description
OpenID OAuth2 source

# Impacts
Authentication for OpenID

# Delete branch after merge
YES 

# NEWS file entries
## New Features
* Added the support for OpenID connect

### CODE REVIEW

File: html/captive-portal/lib/captiveportal/DynamicRouting/Module/Authentication/OAuth/OpenID.pm [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+BEGIN { extends 'captiveportal::PacketFence::DynamicRouting::Module::Authentication::OAuth::OpenID'; }
+
+=head1 NAME
+
+captiveportal::DynamicRouting::Module::Authentication::OAuth::OpenID - OpenID Controller for captiveportal
+
+=head1 DESCRIPTION
+
+[enter your description here]
```
Comment (extrafu) [inline html/captive-portal/lib/captiveportal/DynamicRouting/Module/Authentication/OAuth/OpenID.pm]: Provide it.

File: html/pfappserver/lib/pfappserver/Form/Config/Authentication/Source/OpenID.pm [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+has_field 'redirect_url' =>
+  (
+   type => 'Text',
+   label => 'Portal URL',
+   required => 1,
+   default => pf::Authentication::Source::OpenIDSource->meta->get_attribute('redirect_url')->default,
+   element_attr => {'placeholder' => pf::Authentication::Source::OpenIDSource->meta->get_attribute('redirect_url')->default},
+   element_class => ['input-xlarge'],
+   tags => { after_element => \&help,
+             help => 'The hostname must be the one of your captive portal.' },
```
Comment (extrafu) [inline html/pfappserver/lib/pfappserver/Form/Config/Authentication/Source/OpenID.pm]: That's not clear, rephrase.

File: html/pfappserver/lib/pfappserver/Form/Config/Authentication/Source/OpenID.pm [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  );
+
+has_field 'domains' =>
+  (
+   type => 'Text',
+   label => 'Authorized domains',
+   required => 1,
+   element_class => ['input-xlarge'],
+   tags => { after_element => \&help,
+             help => 'Comma separated list of domains that will be resolve with the correct IP addresses.' },
```
Comment (extrafu) [inline html/pfappserver/lib/pfappserver/Form/Config/Authentication/Source/OpenID.pm]: Comma-separated
resolved

File: html/pfappserver/lib/pfappserver/Form/Config/Authentication/Source/OpenID.pm (hunk 2 above)
Comment (whitx) [reply html/pfappserver/lib/pfappserver/Form/Config/Authentication/Source/OpenID.pm]: This has been the case for every OAuth2 source

File: html/pfappserver/lib/pfappserver/Form/Config/Authentication/Source/OpenID.pm [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    },
+);
+
+has_field 'local_account_logins' => (
+    type => 'PosInteger',
+    label => 'Amount of logins for the local account',
+    default => pf::Authentication::Source::OpenIDSource->meta->get_attribute('local_account_logins')->default,
+    tags => {
+        after_element => \&help_list,
+        help => 'The amount of times, the local account can be used after its created. 0 means infinite.'
```
Comment (extrafu) [inline html/pfappserver/lib/pfappserver/Form/Config/Authentication/Source/OpenID.pm]: The number of times
its -> it is

File: lib/pf/Authentication/Source/OpenIDSource.pm [hunk 5]
```diff
@@ -0,0 +1,71 @@
+package pf::Authentication::Source::OpenIDSource;
+
+=head1 NAME
+
+pf::Authentication::Source::OpenIDSource
+
+=head1 DESCRIPTION
+
+This module implements methods for the OpenID source and the methods necessary
+to perform the OAuth flow since Net::OAuth2 lacks support for OpenID OAuth.
```
Comment (julsemaan) [inline lib/pf/Authentication/Source/OpenIDSource.pm]: Hum, that's not true, seems like a copy/paste from the Twitter source but this doesn't apply.
Use a description similar to Facebook, Google, Github, etc

**Comment by julsemaan** [comment]
Also, @jrouzierinverse, this collides with your work in #2935

Once @whitx fixes the little :nail_care: issue, its ready to merge

Do you prefer integrating his work + commits in #2935 or we merge this and you rebase #2935 on devel post-merge ?

**Comment by jrouzierinverse** [comment]
It would be better for this to hit first then I would rebase

**Pull request merged**